chapter/methodology.md

```diff
@@ -0,0 +1,964 @@
+pub trait LinearizationState {}
+
+pub struct Linearization<S: LinearizationState> {
+    pub state: S,
+}
+```
+
+```{.rust #lst:nls-definition-building-type caption="Type Definition of Building state"}
+pub struct Building {
+    pub linearization: Vec<LinearizationItem<Unresolved>>,
+    pub scope: HashMap<Vec<ScopeId>, Vec<ID>>,
+}
+impl LinearizationState for Building {}
+```
+
+```{.rust #lst:nls-definition-completed-type caption="Type Definition of Completed state" }
+pub struct Completed {
+    pub linearization: Vec<LinearizationItem<Resolved>>,
+    scope: HashMap<Vec<ScopeId>, Vec<ID>>,
+    id_to_index: HashMap<ID, usize>,
+}
+impl LinearizationState for Completed {}
+```
+
+### Transfer from AST
+
+The NLS project aims to present a transferable architecture that can be adapted for future languages.
+Consequently, NLS faces the challenge of satisfying multiple goals
+
+1. To keep up with the frequent changes to the Nickel language and ensure compatibility at minimal cost, NLS needs to *integrate critical functions* of Nickel's runtime
+2. Adaptions to Nickel to accommodate the language server should be minimal not obstruct its development and *maintain performance of the runtime*.
+<!-- what is more? -->
+
+To accommodate these goals NLS comprises three different parts as shown in [@fig:nls-nickel-structure].
+
+The `Linearizer` trait 
+  ~ acts as an interface between Nickel and the language server.
+    NLS implements a `Linearizer` specialized to Nickel which registers AST nodes and builds a final linearization.
+Nickel's type checking implementation
+  ~ was adapted to pass AST nodes to the `Linearizer`.
+    Modifications to Nickel are minimal, comprising only few additional function calls and a slightly extended argument list.
+A stub implementation
+  ~ of the `Linearizer` trait is used during normal operation.
+    Since most methods of this implementation are `no-op`s, the compiler should be able to optimize away all `Linearizer` calls in release builds.
+
+<!-- TODO: caption -->
+```{.graphviz #fig:nls-nickel-structure caption="Interaction of Componenets"}
+digraph {
+  nls [label="NLS"]
+  nickel [label="Nickel"]
+  als [label="Linearizer", shape=box]
+  stub [label="Stub interface"]
+
+  nls ->  nickel  [label="uses"]
+  nls ->  als [label="implements"]
+  stub ->  als [label="implements"]
+  nickel ->  als  [label="uses"]
+  nickel ->  stub  [label="uses"]
+}
+```
+
+
+#### Usage Graph
+
+At the core the linearization is a simple *linear* structure.
+Yet, it represents relationships of nodes on a structural level as a tree-like structure.
+Taking into account variable usage information adds back-edges to the original AST, yielding a graph structure.
+Both kinds of edges have to be encoded with the elements in the list.
+Alas, items have to be referred to using `id`s since the index of items cannot be relied on(such as in e.g. a binary heap), because the array is reordered to optimize access by source position.
```

Comment:
```suggestion
Alas, items have to be referred to using `id`s since the index of items cannot be relied on (such as in e.g. a binary heap), because the array is reordered to optimize access by source position.
```